introspect/providers.go

```diff
@@ -1,39 +1,54 @@
 package introspect
 
-// ConnID represents a connection ID string
-type ConnID string
+import introspect_pb "github.com/libp2p/go-libp2p-core/introspect/pb"
 
-// QueryOutputType determines the way a output needs to be represented in a query result
-type QueryOutputType int
+type (
+	// QueryOutput determines the output form of a query result.
+	QueryOutput int
+
+	// ConnectionID represents a connection ID.
+	ConnectionID string
+
+	// StreamID represents a stream ID.
+	StreamID string
+)
 
 const (
-	// QueryOutputTypeFull dictates that we need to resolve the whole object in the query output
-	QueryOutputTypeFull QueryOutputType = iota
-	// QueryOutputTypeIds dictates that we need to resolve only the identifiers of the object in the query output
-	QueryOutputTypeIds
+	// QueryOutputFull dictates that we need to resolve the whole object in the
+	// query output.
+	QueryOutputFull QueryOutput = iota
+
+	// QueryOutputList dictates that we need to resolve only the identifiers of
+	// the object in the query output.
+	QueryOutputList
 )
 
-// ProvidersMap is a struct which provider modules use to register their interest in providing
-// data to be recorded in a time slice of the running system.
-type ProvidersMap struct {
-	Runtime    func() (*Runtime, error)
-	Connection func(ConnectionQueryInput) ([]*Connection, error)
-	Traffic    func() (*Traffic, error)
-}
+// EXPERIMENTAL. DataProviders enumerates the functions that resolve each entity
+// type. It is used by go-libp2p modules to register callback functions capable
+// of processing entity queries.
+type DataProviders struct {
+	// Runtime is the provider function that returns system runtime information.
+	Runtime func() (*introspect_pb.Runtime, error)
 
-// ConnListQueryType is an Enum to represent the types of queries that can be made when looking up streams
-type ConnListQueryType int
+	// Connection is the provider that is called when information about
+	// Connections is required.
+	Connection func(ConnectionQueryParams) ([]*introspect_pb.Connection, error)
 
-const (
-	// ConnListQueryTypeAll represents a query to fetch all connections
-	ConnListQueryTypeAll ConnListQueryType = iota
-	// ConnListQueryTypeForIds represents a query to fetch connections for a given set of Ids
-	ConnListQueryTypeForIds
-)
+	// Stream is the provider that is called when information about Streams is
+	// required.
+	Stream func(StreamQueryParams) (introspect_pb.StreamList, error)
+
+	// Traffic is the provider that is called when information about network
+	// statistics is required.
+	Traffic func() (*introspect_pb.Traffic, error)
+}
+
+type ConnectionQueryParams struct {
+	Output  QueryOutput
```

Comment:
@raulk  This struct is not enough to represent all the params of a connection query. 

How would you specify whether you want ONLY Ids vs full objects for the Streams in the connection ? The `Output` field here specifies whether we want full connection objects vs connection Ids. But, what about the streams ?